wallet/src/main/scala/org/bitcoins/wallet/api/CoinSelector.scala

```diff
@@ -0,0 +1,69 @@
+package org.bitcoins.wallet.api
+
+import org.bitcoins.core.currency.{CurrencyUnit, CurrencyUnits}
+import org.bitcoins.core.protocol.transaction.TransactionOutput
+import org.bitcoins.core.wallet.fee.FeeUnit
+import org.bitcoins.wallet.models.UTXOSpendingInfoDb
+
+import scala.annotation.tailrec
+
+/** Implements algorithms for selecting from a UTXO set to spend to an output set at a given fee rate. */
+trait CoinSelector {
+
+  /** Greedily selects from walletUtxos starting with the largest outputs, skipping detrimental outputs */
+  def accumulateLargest(
+      walletUtxos: Seq[UTXOSpendingInfoDb],
+      outputs: Seq[TransactionOutput],
+      feeRate: FeeUnit): Vector[UTXOSpendingInfoDb] = {
+    val sortedUtxos =
+      walletUtxos.toVector.sortBy(_.value.satoshis.toLong).reverse
+
+    accumulate(sortedUtxos, outputs, feeRate)
+  }
+
+  /** Greedily selects from walletUtxos starting with the smallest outputs, skipping detrimental outputs */
+  def accumulateSmallestViable(
+      walletUtxos: Seq[UTXOSpendingInfoDb],
+      outputs: Seq[TransactionOutput],
+      feeRate: FeeUnit): Vector[UTXOSpendingInfoDb] = {
+    val sortedUtxos = walletUtxos.toVector.sortBy(_.value.satoshis.toLong)
+
+    accumulate(sortedUtxos, outputs, feeRate)
+  }
+
+  /** Greedily selects from walletUtxos in order, skipping detrimental outputs */
+  def accumulate(
+      walletUtxos: Seq[UTXOSpendingInfoDb],
+      outputs: Seq[TransactionOutput],
+      feeRate: FeeUnit): Vector[UTXOSpendingInfoDb] = {
+    val totalValue = outputs.foldLeft(CurrencyUnits.zero) {
+      case (totVal, output) => totVal + output.value
+    }
+
+    @tailrec
+    def addUtxos(
+        alreadyAdded: Vector[UTXOSpendingInfoDb],
+        valueSoFar: CurrencyUnit,
+        utxosLeft: Vector[UTXOSpendingInfoDb]): Vector[UTXOSpendingInfoDb] = {
+      val fee = CurrencyUnits.zero // TODO: Calculate this
```

Comment:
This is tricky, and I was going to perhaps we should assume that the user calculates this _before_ calling anything in the coin selection algorithm. However this would create a recursive dependency between the [`TxBuilder.creditingOutputs`](https://github.com/bitcoin-s/bitcoin-s-core/blob/master/core/src/main/scala/org/bitcoins/core/wallet/builder/TxBuilder.scala#L74) and the coin selection algorithm. 

I think the right path is 
1. Call coin selection with a `target: CurrencyUnit` 
2. Pass in a desired `FeeUnit` 
3. Call `TxBuilder` with utxo combinations and use [`TxBuilder.unsignedTx`](https://github.com/bitcoin-s/bitcoin-s-core/blob/master/core/src/main/scala/org/bitcoins/core/wallet/builder/TxBuilder.scala#L103-L110) to get the _actual fee_ (not just the sat/byte!) **
4. If we have enough money with this utxo combination, return it. Else iterate again.

** there is a complication here, there isn't an _easy_ way to get the actual fee from the `unsignedTx` returned by `TxBuilder.unsignedTx`. We should consider modifying `TxBuilder` to provide this information to us some how.  I guess if we also want to construct a `TxBuilder`, we need to pass in all relevant information needed to the construct the TxBuilder, which seems unfortunate.

Reply:
That would mean the change output needs to be provided in this method, then? 

I agree on the last point. Would also make sense for the tx builder to inform which output is the change output, IMO. Changing the tx builder should wait until a certain closed source project is merged into upstream, however...